### ISSUE TITLE
Batched MutationObserver callbacks make it possibly difficult to reduce changes.

### ISSUE THREAD

**Issue by trusktr** [whatwg/dom#484]
Suppose we use MutationObserver with `subtree: true` on a root node of a DOM tree.

Suppose within one synchronous tick many changes happen (thousands of changes, subtrees added and removed).

There will be many records.

How do we deduce from those records what the net changes are?

If we make a naive loop that reacts to every change, then we're back at square one: performance of the deprecated mutation events.

We have to instead analyze the mutation record list to determine the net changes. For example, after thousands of changes, it is possible (despite being unlikely) that the net change is a single node was removed. We may want to only respond to that single net change.

But the task of analyzing the changeset is [very difficult](https://stackoverflow.com/questions/38998599).

In the end, the easiest approach (assuming many many changes can happen, we just don't know what end users will do if we are library authors) is to just use a dom-diffing library to compare the previous tree with the new tree, completely ignoring the mutation record list.

This will give us the net changes that we should react to.

Another approach is to not use `subtree: true`, only `childList: true`, and observe every node in the tree. This can possibly be more expensive than the whole-tree diffing approach because it may fire reaction to more changes, but it can still be faster than simply reacting to every change.

---

Any implementation may need to, for example, keep track of a given node to see if the given node is simply moved from one place to another place, and react to a single disconnected change, and a single connected change. The original mutation record list might reveal that the given node was connected and disconnected many times, although the net change by the time the observer's callback fired was that the node was disconnected, then reconnected, and we don't care to run logic for every time the node was connected/disconnected, only reactions for the ultimate net change.

This is needed, for example, when rendering WebGL based on custom elements (for example, a-frame). We only care about what parts of the DOM tree were modified by the time the next animation frame happens, so that we can efficiently determine what to render.

---

Basically, do you have any recommendation on how to handle such analysis and reduction of the changes we observe in the mutation record list of an observer's callback?

**Comment by annevk** [comment]
https://github.com/rafaelw/mutation-summary is a library of sorts that accomplishes this. It doesn't appear to be maintained anymore though. I'm not aware of any new efforts in this area.

**Comment by annevk** [comment]
Closing this since for now we'll leave this to libraries. If there's ever widespread high-level patterns that are worth exposing somehow for performance reasons we'll reconsider.

**Comment by trusktr** [comment]
In case anyone stumbles here, I've been batched mutations [like this]9https://github.com/lume/lume/blob/32d368569af870e74ebdfac5e45f04f58466a961/src/core/Utility.ts#L16-L79)

Here's the code snippet from there:

```js
// TODO padd an options object to make it more clear what the args are.
function observeChildren(target: any, onConnect: any, onDisconnect: any, skipTextNodes: any) {
	// TODO this Map is never cleaned, leaks memory. Maybe use WeakMap
	const childObserver = createChildObserver(onConnect, onDisconnect, skipTextNodes)
	childObserver.observe(target, {childList: true})
	return childObserver
}

// NOTE: If a child is disconnected then connected to the same parent in the
// same turn, then the onConnect and onDisconnect callbacks won't be called
// because the DOM tree will be back in the exact state as before (this is
// possible thanks to the logic associated with weightsPerTarget).
function createChildObserver(onConnect: any, onDisconnect: any, skipTextNodes = false) {
	return new MutationObserver(changes => {
		const weightsPerTarget = new Map<any, Map<any, any>>()

		// We're just counting how many times each child node was added and
		// removed from the parent we're observing.
		for (let i = 0, l = changes.length; i < l; i += 1) {
			const change = changes[i]

			if (change.type != 'childList') continue

			let weights = weightsPerTarget.get(change.target)

			if (!weights) weightsPerTarget.set(change.target, (weights = new Map()))

			const {addedNodes} = change
			for (let l = addedNodes.length, i = 0; i < l; i += 1)
				weights.set(addedNodes[i], (weights.get(addedNodes[i]) || 0) + 1)

			const {removedNodes} = change
			for (let l = removedNodes.length, i = 0; i < l; i += 1)
				weights.set(removedNodes[i], (weights.get(removedNodes[i]) || 0) - 1)
		}

		// NOTE, the destructuring inside the for..of header currently doesn't
		// work due to a Buble bug, so we destructure inside the loop instead.
		// https://github.com/Rich-Harris/buble/issues/182
		// for (const [target, weights] of Array.from(weightsPerTarget)) {
		for (const entry of Array.from(weightsPerTarget)) {
			const [target, weights] = entry

			// for (const [node, weight] of Array.from(weights)) {
			for (const entry of Array.from(weights)) {
				const [node, weight] = entry

				if (skipTextNodes && (node instanceof Text || node instanceof Comment)) continue

				// If the number of times a child was added is greater than the
				// number of times it was removed, then the net result is that
				// it was added, so we call onConnect just once.
				if (weight > 0 && typeof onConnect == 'function') onConnect.call(target, node)
				// If the number of times a child was added is less than the
				// number of times it was removed, then the net result is that
				// it was removed, so we call onDisconnect just once.
				else if (weight < 0 && typeof onDisconnect == 'function') onDisconnect.call(target, node)

				// If the number of times a child was added is equal to the
				// number of times it was removed, then it was essentially left
				// in place, so we don't call anything.
			}
		}
	})
}
```
